inductor: fallback ConvTranspose when output_padding is big

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100846


Fixes https://github.com/pytorch/pytorch/issues/100225.

Do not use mkldnn when `output_padding` is big to align the behavior with eager mode:
https://github.com/pytorch/pytorch/blob/7d0e4e2aa843ef5d73646f5b304914d2b65db93c/aten/src/ATen/native/Convolution.cpp#L500-L507

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

chunyuan-w committed:
inductor: fallback ConvTranspose when output_padding is big

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fallback ConvTranspose when output_padding is big"



Fixes https://github.com/pytorch/pytorch/issues/100225.

Do not use mkldnn when `output_padding` is big to follow the same logic as eager mode:
https://github.com/pytorch/pytorch/blob/7d0e4e2aa843ef5d73646f5b304914d2b65db93c/aten/src/ATen/native/Convolution.cpp#L500-L507

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fallback ConvTranspose when output_padding is big"



Fixes https://github.com/pytorch/pytorch/issues/100225.

Do not use mkldnn when `output_padding` is big to follow the same logic as eager mode:
https://github.com/pytorch/pytorch/blob/7d0e4e2aa843ef5d73646f5b304914d2b65db93c/aten/src/ATen/native/Convolution.cpp#L500-L507

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fallback ConvTranspose when output_padding is big"



Fixes https://github.com/pytorch/pytorch/issues/100225.

Do not use mkldnn when `output_padding` is big to follow the same logic as eager mode:
https://github.com/pytorch/pytorch/blob/7d0e4e2aa843ef5d73646f5b304914d2b65db93c/aten/src/ATen/native/Convolution.cpp#L500-L507

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/mkldnn.py:
```diff
@@ -20,6 +20,14 @@ def is_group_depthwise_conv_transpose(m):
 
 
+# Port from: aten/src/ATen/native/Convolution.cpp:is_output_padding_big
+def is_output_padding_big(m):
+    is_big = False
+    for i in range(len(m.output_padding)):
+        is_big |= m.output_padding[i] >= m.stride[i]
+    return is_big
```
a more pythonic impl and you don't need an extra function:
`any(padding >= stride for padding, stride in zip(m.output_padding, m.stride))`

chunyuan-w committed:
Update on "inductor: fallback ConvTranspose when output_padding is big"



Fixes https://github.com/pytorch/pytorch/issues/100225.

Do not use mkldnn when `output_padding` is big to align the behavior with eager mode:
https://github.com/pytorch/pytorch/blob/7d0e4e2aa843ef5d73646f5b304914d2b65db93c/aten/src/ATen/native/Convolution.cpp#L500-L507

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor: fallback ConvTranspose when output_padding is big"



Fixes https://github.com/pytorch/pytorch/issues/100225.

Do not use mkldnn when `output_padding` is big to align the behavior with eager mode:
https://github.com/pytorch/pytorch/blob/7d0e4e2aa843ef5d73646f5b304914d2b65db93c/aten/src/ATen/native/Convolution.cpp#L500-L507

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on torch/_inductor/mkldnn.py:
```diff
@@ -314,6 +314,12 @@ def pack_module(gm: torch.fx.GraphModule):
                     or len(node.args) > 1
                     or len(node.kwargs) > 0
+                    or any(
+                        output_padding >= stride
```
This won't work with dynamic shapes.  You will just get `TypeError: cannot determine truth value of Relational`
```
>>> 10 > sympy.Symbol("x")
x < 10
```

chunyuan-w committed:
Update on "inductor: fallback ConvTranspose when output_padding is big"



Fixes https://github.com/pytorch/pytorch/issues/100225.

Do not use mkldnn when `output_padding` is big to align the behavior with eager mode:
https://github.com/pytorch/pytorch/blob/7d0e4e2aa843ef5d73646f5b304914d2b65db93c/aten/src/ATen/native/Convolution.cpp#L500-L507

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge
